File: src/cod_cookie_jar/adapters/cdp.py
"""Primary adapter: raw Chrome DevTools Protocol.

Talks to a browser you launched yourself with, e.g.:
    chrome --remote-debugging-port=9222

Uses ``Storage.getCookies`` (the non-deprecated successor to
``Network.getAllCookies``), which returns the full cookie store including
HttpOnly cookies, with no on-disk decryption and no extension loading.

This adapter only reads from a debugging endpoint the operator explicitly
enabled on their own browser. It performs no on-disk access and no bypass of
any browser encryption.
"""
from __future__ import annotations

import json
import urllib.request
from typing import Optional

from ..core import NormalizedCookie
from .base import Adapter, filter_domain


class CdpAdapter(Adapter):
    def __init__(self, endpoint: str = "http://localhost:9222", timeout: float = 10.0):
        self.endpoint = endpoint.rstrip("/")
        self.timeout = timeout

    def _browser_ws_url(self) -> str:
        with urllib.request.urlopen(
            self.endpoint + "/json/version", timeout=self.timeout
        ) as r:
            info = json.loads(r.read().decode("utf-8"))
        url = info.get("webSocketDebuggerUrl")
        if not url:
            raise RuntimeError(
                "No webSocketDebuggerUrl at %s/json/version -- is the browser "
                "running with --remote-debugging-port?" % self.endpoint
            )
        return url

    def _raw_cookies(self) -> list[dict]:
        try:
            import websocket  # from the 'websocket-client' package
        except ImportError as e:  # pragma: no cover
            raise RuntimeError(
                "The CDP adapter needs 'websocket-client'. "
                "Install with:  pip install cod-cookie-jar[cdp]"
            ) from e

        ws = websocket.create_connection(
            self._browser_ws_url(), timeout=self.timeout
        )
        try:
            ws.send(json.dumps({"id": 1, "method": "Storage.getCookies"}))
            while True:
                msg = json.loads(ws.recv())
                if msg.get("id") == 1:
                    if "error" in msg:
                        raise RuntimeError("CDP error: %s" % msg["error"])
                    return msg["result"]["cookies"]
        finally:
            ws.close()

    def fetch(self, domain: Optional[str] = None) -> list[NormalizedCookie]:
        cookies = []
        for c in self._raw_cookies():
            session = bool(c.get("session")) or c.get("expires", -1) in (-1, None)
            cookies.append(
                NormalizedCookie(
                    name=c["name"],
                    value=c["value"],
                    domain=c["domain"],
                    path=c.get("path", "/"),
                    secure=bool(c.get("secure")),
                    http_only=bool(c.get("httpOnly")),
                    expires=None if session else float(c["expires"]),
                    same_site=c.get("sameSite"),
                    partition_key=(
                        c.get("partitionKey", {}).get("topLevelSite")
                        if isinstance(c.get("partitionKey"), dict)
                        else c.get("partitionKey")
                    ),
                )
            )
        return filter_domain(cookies, domain)

File: node/src/adapters/cdp.js
/**
 * CDP adapter for the Node port.
 *
 * Talks to a browser YOU launched with --remote-debugging-port=9222, using
 * Storage.getCookies over the browser websocket. No dependencies: Node 18+ has
 * a global WebSocket and global fetch.
 *
 * Reads only from a debugging endpoint the operator explicitly enabled on their
 * own browser. No on-disk access, no encryption bypass.
 */
import { filterDomain } from './core.js';

async function browserWsUrl(endpoint, timeoutMs) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(endpoint.replace(/\/$/, '') + '/json/version', {
      signal: ctrl.signal,
    });
    const info = await res.json();
    if (!info.webSocketDebuggerUrl) {
      throw new Error(
        `No webSocketDebuggerUrl at ${endpoint}/json/version -- is the browser ` +
          'running with --remote-debugging-port?',
      );
    }
    return info.webSocketDebuggerUrl;
  } finally {
    clearTimeout(t);
  }
}

function rawCookies(wsUrl, timeoutMs) {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(wsUrl);
    const timer = setTimeout(() => {
      ws.close();
      reject(new Error('CDP timeout'));
    }, timeoutMs);

    ws.addEventListener('open', () => {
      ws.send(JSON.stringify({ id: 1, method: 'Storage.getCookies' }));
    });
    ws.addEventListener('message', (ev) => {
      const msg = JSON.parse(ev.data);
      if (msg.id === 1) {
        clearTimeout(timer);
        ws.close();
        if (msg.error) reject(new Error('CDP error: ' + JSON.stringify(msg.error)));
        else resolve(msg.result.cookies);
      }
    });
    ws.addEventListener('error', (e) => {
      clearTimeout(timer);
      reject(new Error('WebSocket error: ' + (e.message || 'connection failed')));
    });
  });
}

export class CdpAdapter {
  constructor(endpoint = 'http://localhost:9222', timeoutMs = 10000) {
    this.endpoint = endpoint;
    this.timeoutMs = timeoutMs;
  }

  async fetch(domain = null) {
    const wsUrl = await browserWsUrl(this.endpoint, this.timeoutMs);
    const raw = await rawCookies(wsUrl, this.timeoutMs);
    const cookies = raw.map((c) => {
      const session = Boolean(c.session) || c.expires === undefined || c.expires < 0;
      let partitionKey = null;
      if (c.partitionKey) {
        partitionKey =
          typeof c.partitionKey === 'object'
            ? c.partitionKey.topLevelSite
            : c.partitionKey;
      }
      return {
        name: c.name,
        value: c.value,
        domain: c.domain,
        path: c.path || '/',
        secure: Boolean(c.secure),
        httpOnly: Boolean(c.httpOnly),
        expires: session ? null : Number(c.expires),
        hostOnly: null,
        sameSite: c.sameSite ?? null,
        partitionKey,
      };
    });
    return filterDomain(cookies, domain);
  }
}
